File: samples/objective-c/backendlessDemos/PhotoStreamer/ViewController.m
//
//  ViewController.m
//  PhotoStreamer
//
//  Created by Slava Vdovichenko on 7/6/15.
//  Copyright (c) 2015 BACKENDLESS.COM. All rights reserved.
//

#import "ViewController.h"
#import <mach/mach_time.h>
#import "Backendless.h"
#import "MediaService.h"

//#define VIDEO_TUBE @"videoTube"
#define VIDEO_TUBE @"Default"
#define DEFAULT_STREAM_NAME @"photo"
#define DEFAULT_FPS 10


@interface ViewController () <IMediaStreamerDelegate> {
    MediaPublisher *_publisher;
}

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    @try {
        
        [backendless initAppFault];
    }
    @catch (Fault *fault) {
        NSLog(@"initAppFault -> %@", fault);
        [self showAlert:fault.message];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark -
#pragma mark Private Methods

-(void)showAlert:(NSString *)message {
    UIAlertView *av = [[UIAlertView alloc] initWithTitle:@"Error:" message:message delegate:nil cancelButtonTitle:@"Ok" otherButtonTitles:nil];
    [av show];
}

-(int64_t)getTimestampMs {
    mach_timebase_info_data_t info;
    mach_timebase_info(&info);
    return 1e-6*mach_absolute_time()*info.numer/info.denom;
}

-(void)serialImage:(UIImage *)image times:(int)times {
    
    if (!_publisher)
        return;
    
    int64_t _timestamp = [self getTimestampMs];
    
    NSLog(@"serialImage: timestamp = %lld", _timestamp);
    
    [_publisher sendImage:[image CGImage] timestamp:_timestamp];
    
    if (!times--)
        return;

    dispatch_time_t interval = dispatch_time(DISPATCH_TIME_NOW, 1ull*NSEC_PER_MSEC*1000/DEFAULT_FPS);
    dispatch_after(interval, dispatch_get_main_queue(), ^{
        [self serialImage:image times:times];
    });

}

#pragma mark -
#pragma mark IBAction

-(IBAction)publish:(id)sender {
    
    if (_publisher)
        return;
    
    MediaPublishOptions *options = [MediaPublishOptions liveStream:nil];
    options.orientation = AVCaptureVideoOrientationPortrait;
    
    [options setCustomVideo:DEFAULT_FPS width:640 height:640];
   
    _publisher =[backendless.mediaService publishStream:DEFAULT_STREAM_NAME tube:VIDEO_TUBE options:options responder:self];
    
    NSLog(@"publishControl: -> published stream: '%@/%@'", _publisher.streamPath, _publisher.streamName);
    
    self.btnPublish.hidden = YES;
    
    [self.netActivity startAnimating];
}

-(IBAction)stop:(id)sender {
    
    if (_publisher)
    {
        [_publisher disconnect];
        _publisher = nil;
        
        self.btnPublish.hidden = NO;
        self.btnStop.hidden = YES;
        self.btnTakePhoto.hidden = YES;
        self.imageView.hidden = YES;
        
        [self.netActivity stopAnimating];
    }
}

-(IBAction)takePhoto:(id)sender {
    
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.allowsEditing = YES;
    picker.sourceType = UIImagePickerControllerSourceTypeCamera;
    
    [self presentViewController:picker animated:YES completion:NULL];
}

#pragma mark -
#pragma mark - Image Picker Controller delegate methods

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary *)info {
    
    UIImage *image = info[UIImagePickerControllerEditedImage];
    [self.imageView performSelectorOnMainThread:@selector(setImage:) withObject:image waitUntilDone:NO]; // MAIN THREAD
    [self serialImage:image times:10];
    
    [picker dismissViewControllerAnimated:YES completion:NULL];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    
    [picker dismissViewControllerAnimated:YES completion:NULL];
}

#pragma mark -
#pragma mark IMediaStreamerDelegate Methods

-(void)streamStateChanged:(id)sender state:(int)state description:(NSString *)description {
    
    NSLog(@"<IMediaStreamerDelegate> streamStateChanged: %d = %@", (int)state, description);
    
    switch (state) {
            
        case CONN_DISCONNECTED: {
            
            if ([description isEqualToString:MP_STREAM_IS_BUSY]) {
                [self showAlert:[NSString stringWithString:description]];
            }
            
            [self stop:sender];
            break;
        }
            
        case CONN_CONNECTED: {
            break;
        }
            
        case STREAM_CREATED: {
            break;
        }
            
        case STREAM_PLAYING: {
            
            if (!([description isEqualToString:MP_NETSTREAM_PUBLISH_START] || [description isEqualToString:MP_RTMP_CLIENT_STREAM_IS_PLAYING])) {
                [self stop:sender];
                break;
            }
            
            self.btnStop.hidden = NO;
            self.btnTakePhoto.hidden = NO;
            self.imageView.hidden = NO;
            
            [self.netActivity stopAnimating];
           
            break;
        }
            
        case STREAM_PAUSED: {
            
            if ([description isEqualToString:MP_RTMP_CLIENT_STREAM_IS_PAUSED]) {
                break;
            }
            
            if ([description isEqualToString:MP_NETSTREAM_PLAY_STREAM_NOT_FOUND]) {
                [self showAlert:[NSString stringWithString:description]];
            }
            
            [self stop:sender];
            break;
        }
            
        default:
            break;
    }
}

-(void)streamConnectFailed:(id)sender code:(int)code description:(NSString *)description {
    
    NSLog(@"<IMediaStreamerDelegate> streamConnectFailed: %d = %@", code, description);
    
    [self stop:sender];
    
    [self showAlert:(code == -1) ?
     [NSString stringWithFormat:@"Unable to connect to the server. Make sure the hostname/IP address and port number are valid\n"] :
     [NSString stringWithFormat:@"connectFailedEvent: %@ \n", description]];
}

@end

File: VideoService/VideoService/ViewController.swift
//
//  ViewController.swift
//  VideoService
//
//  Created by Vyacheslav Vdovichenko on 10/24/14.
//  Copyright (c) 2014 BACKENDLESS.COM. All rights reserved.
//

import UIKit
//import AVFoundation

class ViewController: UIViewController, UITextFieldDelegate/*, IMediaStreamerDelegate*/ {

    @IBOutlet var btnPublish : UIButton!
    @IBOutlet var btnPlayback : UIButton!
    @IBOutlet var btnStopMedia : UIButton!
    @IBOutlet var btnSwapCamera : UIButton!
    @IBOutlet var preView : UIView!
    @IBOutlet var playbackView : UIImageView!
    @IBOutlet var textField : UITextField!
    @IBOutlet var lblLive : UILabel!
    @IBOutlet var switchView : UISwitch!
    @IBOutlet var netActivity : UIActivityIndicatorView!
    
    let VIDEO_TUBE = "videoTube"
    
    var backendless = Backendless.sharedInstance()
    
    var _publisher: MediaPublisher?
    var _player: MediaPlayer?

    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func switchCamerasControl(sender: AnyObject) {
        
        println("----------------------- pswitchCamerasControl ------------------------------------------------------")
        
        _publisher?.switchCameras()
    }
    
    @IBAction func stopMediaControl(sender: AnyObject) {
        
        println("----------------------- stopMediaControl ------------------------------------------------------")
        
        if (_publisher != nil) {
            
            _publisher?.disconnect()
            _publisher = nil;
            
            self.preView.hidden = true
            self.btnStopMedia.hidden = true
            self.btnSwapCamera.hidden = true
        }
        
        if (_player != nil)
        {
            _player?.disconnect()
            _player = nil;
            self.playbackView.hidden = true
            self.btnStopMedia.hidden = true
        }
        
        self.btnPublish.hidden = false
        self.btnPlayback.hidden = false
        self.textField.hidden = false
        self.switchView.hidden = false
        self.lblLive.hidden = false
        
        self.netActivity.stopAnimating()
    }
    
    @IBAction func playbackControl(sender: AnyObject) {
        
        println("----------------------- playbackControl ------------------------------------------------------")
        
        var options: MediaPlaybackOptions
        if (switchView.on) {
            options = MediaPlaybackOptions.liveStream(self.playbackView) as MediaPlaybackOptions
        }
        else {
            options = MediaPlaybackOptions.recordStream(self.playbackView) as MediaPlaybackOptions
        }
        
        options.orientation = UIImageOrientation(rawValue: 0)! //UIImageOrientationUp
        options.isRealTime = switchView.on
        
        _player = backendless.mediaService.playbackStream(textField.text, tube:VIDEO_TUBE, options:options, responder:nil)
        
        self.btnPublish.hidden = true
        self.btnPlayback.hidden = true
        self.textField.hidden = true
        self.switchView.hidden = true
        self.lblLive.hidden = true
        
        //netActivity.startAnimating()
        
        // -- TEMPOPARY !!!
        self.playbackView.hidden = false
        self.btnStopMedia.hidden = false
        //
    
    }
    
    @IBAction func publishControl(sender: AnyObject) {
        
        println("----------------------- publishControl ------------------------------------------------------")
        
        var options: MediaPublishOptions
        if (switchView.on) {
            options = MediaPublishOptions.liveStream(self.preView) as MediaPublishOptions
        }
        else {
            options = MediaPublishOptions.recordStream(self.preView) as MediaPublishOptions
        }
        
        options.orientation = AVCaptureVideoOrientation(rawValue: 1)! //AVCaptureVideoOrientationPortrait
        options.resolution = RESOLUTION_CIF
        
        _publisher = backendless.mediaService.publishStream(textField.text, tube:VIDEO_TUBE, options:options, responder:nil)
        
        self.btnPublish.hidden = true
        self.btnPlayback.hidden = true
        self.textField.hidden = true
        self.switchView.hidden = true
        self.lblLive.hidden = true
        
        //netActivity.startAnimating()
        
        // -- TEMPOPARY !!!
        self.preView.hidden = false
        self.btnStopMedia.hidden = false
        self.btnSwapCamera.hidden = false
        //

    }
    
    @IBAction func viewTapped(sender: AnyObject) {
        textField.resignFirstResponder()
    }

    // UITextFieldDelegate protocol methods
    
    func textFieldShouldReturn(_textField: UITextField) {
        textField.resignFirstResponder()
    }
    
    // MediaStreamerDelegate  protocol methods
    
    func streamStateChanged(sender: AnyObject, state: MPMediaStreamState, description: NSString) {
    }
    
    func streamConnectFailed(sender: AnyObject, code: Int, description: NSString) {
    }
}

